dsl/src/commonMain/kotlin/com/freeletics/flowredux/dsl/FlowReduxStateMachine.kt

```diff
@@ -1,56 +1,69 @@
 package com.freeletics.flowredux.dsl
 
 import com.freeletics.flowredux.FlowReduxLogger
-import com.freeletics.mad.statemachine.StateMachine
-import kotlinx.coroutines.CoroutineScope
 import kotlinx.coroutines.ExperimentalCoroutinesApi
 import kotlinx.coroutines.FlowPreview
 import kotlinx.coroutines.channels.Channel
-import kotlinx.coroutines.flow.SharingStarted
-import kotlinx.coroutines.flow.StateFlow
+import kotlinx.coroutines.flow.Flow
 import kotlinx.coroutines.flow.consumeAsFlow
-import kotlinx.coroutines.flow.stateIn
-import kotlinx.coroutines.isActive
 
 @FlowPreview
 @ExperimentalCoroutinesApi
 abstract class FlowReduxStateMachine<S : Any, A : Any>(
-    private val initialState: S,
-    private val scope: CoroutineScope,
+    private val initialStateSupplier: () -> S,
     private val logger: FlowReduxLogger? = null
-) : StateMachine<S, A> {
+) {
 
     private val inputActions = Channel<A>()
-    private var internalState: StateFlow<S>? = null
+    private var specBlock: (FlowReduxStoreBuilder<S, A>.() -> Unit)? = null
+    private var specBlockSet = false
+
+    constructor(
+        initialStateSupplier: () -> S
+    ) : this(
+        logger = null,
+        initialStateSupplier = initialStateSupplier
+    )
+
+    constructor(initialState: S) : this(logger = null, initialState = initialState)
+
+    constructor(
+        initialState: S,
+        logger: FlowReduxLogger?
+    ) : this(logger = logger, initialStateSupplier = { initialState })
 
     protected fun spec(specBlock: FlowReduxStoreBuilder<S, A>.() -> Unit) {
-        if (internalState != null) {
+
+        if (this.specBlock != null)
             throw IllegalStateException(
                 "State machine spec has already been set. " +
-                    "It's only allowed to call spec {...} once."
+                        "It's only allowed to call spec {...} once."
             )
-        }
-        internalState = inputActions
-            .consumeAsFlow()
-            .reduxStore(logger, initialState, specBlock)
```

Comment:
We could still have this here , just without the `stateIn`. The flow and therefore the state machine is still cold. Then we don't need `specBlock` and `specBlockSet` and cleanup for those is not needed